### ISSUE TITLE
Enabling Thread Sanitizer in Xcode causes the app to crash

### ISSUE THREAD

**Issue by appleby** [msintuneappsdk/ms-intune-app-sdk-ios#205]
**Describe the bug:**

Enabling Thread Sanitizer in Xcode causes the [Chatr] sample app to crash. This looks like an Intune SDK bug, so I'm filing this here. Let me know if I should move this to the Chatr issue tracker instead. (Also, apologies if this is a known issue. I didn't find anything relevant searching the release notes or the issue tracker here).

Version info:

* Xcode 12.5.1
* iOS 14.5 simulator
* macOS 11.6 (Big Sur)
* Intel macbook pro
* Chatr commit [ad5c2188](https://github.com/msintuneappsdk/Chatr-Sample-Intune-iOS-App/commit/ad5c2188b6ac0cea87866c56f158ccb838dbe0b0) (current master HEAD)

**To Reproduce**
1. Clone the [Chatr] sample app
2. Open the included Xcode project in Xcode 12.5.1
3. Enable the Thread Sanitizer diagnostic for the chatr scheme. (Product > Scheme > Edit Scheme..., then Run > Diagnostics > [x] Thread Sanitizer)
4. Follow setup steps from the [Chatr] README
5. Run the app from Xcode (Product > Run)
6. Authenticate with your Intune credentials
7. After successful auth, you should be prompted to restart the app. "Your organization is now protecting its data in this app. You need to restart the app to continue." Click OK
8. Go back to Xcode and run the app again
9. App crashes due to infinite recursion between tsan's `wrap_fopen` and an unamed Intune symbol (presumably Intune's own `fopen` wrapper).

[Chatr]: https://github.com/msintuneappsdk/Chatr-Sample-Intune-iOS-App

**Expected behavior:**

Expected the app to not crash. With Thread Sanitizer disabled, the app behaves as expected.

**Screenshots and logs:**

**App logs**

```
chatr(64329,0x11a0fce00) malloc: nano zone abandoned due to inability to preallocate reserved vm space.
(lldb) 
```

**Partial stack trace (see attachments for full backtrace)**

[chatr-tsan-stack-trace.txt](https://github.com/msintuneappsdk/ms-intune-app-sdk-ios/files/7268387/chatr-tsan-stack-trace.txt)

```
Thread 1 Queue : com.apple.main-thread (serial)
#0	0x000000010fc22037 in __tsan::cur_thread() ()
#1	0x000000010fbd8b24 in __tsan::ThreadClock::UpdateCurrentThread(__tsan::DenseSlabAllocCache*, __tsan::SyncClock*) const ()
#2	0x000000010fbd8801 in __tsan::ThreadClock::release(__tsan::DenseSlabAllocCache*, __tsan::SyncClock*) ()
#3	0x000000010fc174ab in __tsan::Release(__tsan::ThreadState*, unsigned long, unsigned long) ()
#4	0x000000010fc20984 in wrap_os_unfair_lock_unlock ()
#5	0x00007fff2019161b in objc_object::sidetable_release(bool, bool) ()
#6	0x0000000110d15d15 in ___lldb_unnamed_symbol1013$$IntuneMAMSwift ()
#7	0x0000000110d18bc5 in ___lldb_unnamed_symbol1048$$IntuneMAMSwift ()
#8	0x0000000110ebf6ac in ___lldb_unnamed_symbol8024$$IntuneMAMSwift ()
#9	0x000000010fbf054f in wrap_fopen ()
#10	0x0000000110ebf70b in ___lldb_unnamed_symbol8024$$IntuneMAMSwift ()
#11	0x000000010fbf054f in wrap_fopen ()
#12	0x0000000110ebf70b in ___lldb_unnamed_symbol8024$$IntuneMAMSwift ()
#13	0x000000010fbf054f in wrap_fopen ()
#14	0x0000000110ebf70b in ___lldb_unnamed_symbol8024$$IntuneMAMSwift ()
#15	0x000000010fbf054f in wrap_fopen ()
#16	0x0000000110ebf70b in ___lldb_unnamed_symbol8024$$IntuneMAMSwift ()
#17	0x000000010fbf054f in wrap_fopen ()
#18	0x0000000110ebf70b in ___lldb_unnamed_symbol8024$$IntuneMAMSwift ()
#19	0x000000010fbf054f in wrap_fopen ()
#20	0x0000000110ebf70b in ___lldb_unnamed_symbol8024$$IntuneMAMSwift ()
#21	0x000000010fbf054f in wrap_fopen ()
... repeated many times...
#115814	0x0000000110ebf70b in ___lldb_unnamed_symbol8024$$IntuneMAMSwift ()
#115815	0x000000010fbf054f in wrap_fopen ()
#115816	0x00007fff202f0116 in _fsi_get_user ()
#115817	0x00007fff202f9915 in search_user_byuid ()
#115818	0x00007fff202f424c in getpwuid ()
#115819	0x000000010fbe7eeb in wrap_getpwuid ()
#115820	0x00007fff204872c1 in _CFUnsandboxedHomeDirectoryForCurrentUser ()
#115821	0x00007fff20487175 in -[_CFXPreferences currentUserHasInvalidHomeDirectory] ()
#115822	0x00007fff2049ca50 in -[CFPrefsPlistSource volatilizeIfInvalidHomeDir] ()
#115823	0x00007fff2049c94c in __62-[CFPrefsPlistSource createRequestNewContentMessageForDaemon:]_block_invoke ()
#115824	0x00007fff2049c8c4 in -[CFPrefsPlistSource createRequestNewContentMessageForDaemon:] ()
#115825	0x00007fff2044a3f5 in -[CFPrefsSearchListSource createRequestNewContentMessageForDaemon:] ()
#115826	0x00007fff2044a8e1 in -[CFPrefsSearchListSource alreadylocked_generationCountFromListOfSources:count:] ()
#115827	0x00007fff204491a7 in -[CFPrefsSearchListSource alreadylocked_getDictionary:] ()
#115828	0x00007fff20448d7e in -[CFPrefsSearchListSource alreadylocked_copyValueForKey:] ()
#115829	0x00007fff203e8a5b in -[CFPrefsSource copyValueForKey:] ()
#115830	0x00007fff2048a6c9 in __76-[_CFXPreferences copyAppValueForKey:identifier:container:configurationURL:]_block_invoke ()
#115831	0x00007fff2044c2d7 in __108-[_CFXPreferences(SearchListAdditions) withSearchListForIdentifier:container:cloudConfigurationURL:perform:]_block_invoke ()
#115832	0x00007fff2044b9d2 in normalizeQuintuplet ()
#115833	0x00007fff2044c130 in -[_CFXPreferences withSearchListForIdentifier:container:cloudConfigurationURL:perform:] ()
#115834	0x00007fff2048a5ad in -[_CFXPreferences copyAppValueForKey:identifier:container:configurationURL:] ()
#115835	0x00007fff2048cf99 in _CFPreferencesCopyAppValueWithContainerAndConfiguration ()
#115836	0x00007fff2086edbb in -[NSUserDefaults(NSUserDefaults) init] ()
#115837	0x00007fff2086f28a in +[NSUserDefaults(NSUserDefaults) standardUserDefaults] ()
#115838	0x0000000110ecc491 in ___lldb_unnamed_symbol8225$$IntuneMAMSwift ()
#115839	0x0000000110e9e677 in ___lldb_unnamed_symbol7392$$IntuneMAMSwift ()
#115840	0x0000000110d58b37 in ___lldb_unnamed_symbol2294$$IntuneMAMSwift ()
#115841	0x0000000110d84d7a in ___lldb_unnamed_symbol3077$$IntuneMAMSwift ()
#115842	0x0000000110ee17c0 in ___lldb_unnamed_symbol8656$$IntuneMAMSwift ()
#115843	0x000000010fc24405 in wrap_dispatch_once ()
#115844	0x0000000110ee1980 in ___lldb_unnamed_symbol8657$$IntuneMAMSwift ()
#115845	0x00007fff20181ff2 in load_images ()
#115846	0x000000010fb07e2c in dyld::notifySingle(dyld_image_states, ImageLoader const*, ImageLoader::InitializerTimingList*) ()
#115847	0x000000010fb16ba5 in ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) ()
#115848	0x000000010fb16b24 in ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) ()
#115849	0x000000010fb14ec7 in ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) ()
#115850	0x000000010fb14f68 in ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) ()
#115851	0x000000010fb0826b in dyld::initializeMainExecutable() ()
#115852	0x000000010fb0cf56 in dyld::_main(macho_header const*, unsigned long, int, char const**, char const**, char const**, unsigned long*) ()
#115853	0x000000010fb071c2 in start_sim ()
#115854	0x000000011a02ec27 in dyld::useSimulatorDyld(int, macho_header const*, char const*, int, char const**, char const**, char const**, unsigned long*, unsigned long*) ()
#115855	0x000000011a02c15e in dyld::_main(macho_header const*, unsigned long, int, char const**, char const**, char const**, unsigned long*) ()
#115856	0x000000011a026224 in dyldbootstrap::start(dyld3::MachOLoaded const*, int, char const**, dyld3::MachOLoaded const*, unsigned long*) ()
#115857	0x000000011a026025 in _dyld_start ()
```

**Smartphone (please complete the following information):**
 - iOS x86_64 simulator
 - OS: iOS 14.5

**Intune App SDK for iOS (please complete the following information):**

- *What version of the Intune SDK are you using? Are you using the latest version?* The [Chatr] repo claims it is using 14.7.1
- *What platform is your app based in (native, Xamarin based, Cordova, etc)?* native
- *For errors post build, does the app launch without being Intune SDK integrated?* Yes
- *Who is the customer?* Xcode developers
- *Do you see a trend with it only being reproduced on a specific device?* I think Tsan is only supported on simulators

**Additional context:**

It looks like the IntuneMAMSwift lib included in the Chatr sample repo has some special handling for the asan library, but not tsan. Maybe tsan can be handled similarly?

```
$ strings IntuneMAMSwift.xcframework/ios-arm64_x86_64-simulator/IntuneMAMSwift.framework/IntuneMAMSwift \
      | grep libclang_rt
libclang_rt.asan_ios_dynamic.dylib
libclang_rt.asan_iossim_dynamic.dylib
```

```
$ otool -tvV IntuneMAMSwift.xcframework/ios-arm64_x86_64-simulator/IntuneMAMSwift.framework/IntuneMAMSwift \
       | grep -C 3 libclang_rt
000000000006430b        testl   %eax, %eax
000000000006430d        je      0x644a4
0000000000064313        movq    %r15, %rdi
0000000000064316        leaq    0x25115c(%rip), %rsi            ## literal pool for: "libclang_rt.asan_ios_dynamic.dylib"
000000000006431d        callq   0x267bca                        ## symbol stub for: _strcmp
0000000000064322        testl   %eax, %eax
0000000000064324        je      0x644a4
--
--
0000000000064322        testl   %eax, %eax
0000000000064324        je      0x644a4
000000000006432a        movq    %r15, %rdi
000000000006432d        leaq    0x251168(%rip), %rsi            ## literal pool for: "libclang_rt.asan_iossim_dynamic.dylib"
0000000000064334        callq   0x267bca                        ## symbol stub for: _strcmp
0000000000064339        testl   %eax, %eax
000000000006433b        je      0x644a4
```

**Comment by vigavadi** [comment]
@appleby Thank you for reaching out. You can file all Chatr related issues [here](https://github.com/msintuneappsdk/Chatr-Sample-Intune-iOS-App/issues) 
Does the Chatr app run fine without enabling the Thread Sanitizer?

**Comment by appleby** [comment]
Hi @vigavadi, thanks for responding. Yes, the Chatr app runs fine without enabling Thread Sanitizer.

I should have mentioned that I'm also able to reproduce this crash using another (proprietary) app integrated with Intune, so I don't think this a Chatr bug.  I just used Chatr as a hopefully easy-to-reproduce example.

From the stacktrace, it looks like something has gone wrong replacing calls to fopen with some symbol in IntuneMAMSwift, hence why I've filed the bug here.

If you still think this is Chatr bug let me know. I'd be happy to try linking IntuneMAMSwift with another simple "hello world" iOS app and see if it reproduces there.
